README.md
# REWAVI

[![Build status](https://ci.appveyor.com/api/projects/status/ffcocigoiclo2ekr/branch/master?svg=true)](https://ci.appveyor.com/project/wieslawsoltes/rewavi/branch/master)

[![Github All Releases](https://img.shields.io/github/downloads/wieslawsoltes/rewavi/total.svg)](https://github.com/wieslawsoltes/rewavi/releases)
[![GitHub Release](https://img.shields.io/github/release/wieslawsoltes/rewavi.svg)](https://github.com/wieslawsoltes/rewavi/releases/latest)
[![Github Releases](https://img.shields.io/github/downloads/wieslawsoltes/rewavi/latest/total.svg)](https://github.com/wieslawsoltes/rewavi/releases)

rewritten/modified WAVI

## Download REWAVI

| Platform           | Character Set  | Version   | Download                                                                                                                |
|--------------------|----------------|-----------|-------------------------------------------------------------------------------------------------------------------------|
| Windows 32-bit     | MBCS           | 0.06      | [rewavi-0.06-Win32.zip](https://github.com/wieslawsoltes/rewavi/releases/download/0.06/rewavi-0.06-Win32.zip)           |
| Windows 64-bit     | MBCS           | 0.06      | [rewavi-0.06-x64.zip](https://github.com/wieslawsoltes/rewavi/releases/download/0.06/rewavi-0.06-x64.zip)               |

## System requirements

Minimum supported Windows version is Windows 7 SP1 or above, recommended is Windows 10 Anniversary Update.

Provided binaries should work under Windows XP Service Pack 3 (SP3) for x86, Windows XP Service Pack 2 (SP2) for x64.

Minimum supported Linux version is Ubutnu 16.10 (using Wine 2.0).

## CI Builds

[Download](https://ci.appveyor.com/project/wieslawsoltes/rewavi/build/artifacts) bleeding edge builds from the CI server.

## About

REWAVI - tool of extract PCM audio stream from avi file.

RESILENCE - tool of generate silence PCM WAV file.

## Usage

* [Usage](https://github.com/wieslawsoltes/rewavi/blob/master/docs/usage.txt)
* [Usage2](https://github.com/wieslawsoltes/rewavi/blob/master/docs/usage2.txt)

## Troubleshooting

To report issues please use [issues tracker](https://github.com/wieslawsoltes/rewavi/issues).

For more informations and help please visit [this forum thread](http://forum.doom9.org/showthread.php?t=175174).

## Sources

Sources are available in the [git source code repository](https://github.com/wieslawsoltes/rewavi/).

## Build

### Prerequisites

```
git clone https://github.com/wieslawsoltes/rewavi.git
cd rewavi
git submodule update --init --recursive
```

### Install Visual Studio 2017

* [VS 2017](https://www.visualstudio.com/pl/downloads/)

### Windows 7 SDK

For Windows XP compatibility program is compiled using `Platform Toolset` for `Visual Studio 2017 - Windows XP (v141_xp)`.

For more details please read [Configuring Programs for Windows XP](https://msdn.microsoft.com/en-us/library/jj851139.aspx).

### Build Solution
```
Open rewavi.sln in Visual Studios 2017 or above.
```

## Copyrights

This is fork of [rewavi](https://github.com/chikuzen/rewavi) which is rewritten/modified [WAVI](http://sourceforge.net/projects/wavi-avi2wav/).

### REWAVI

```
REWAVI - rewritten/modified WAVI
(c) 2011 Oka Motofumi
(c) 2018 Wieslaw Soltes
WAVI original - (c) 2007 Tamas Kurucsai, with lots of help from tebasuna51
Licensed under the terms of the GNU General Public License version2 or later.
```

### RESILENCE

```
RESILENCE - rewritten/modified SILENCE
(c) 2011 Oka Motofumi
(c) 2018 Wieslaw Soltes
original SILENCE (c) 2007 Tamas Kurucsai
Licensed under the terms of the GNU General Public License verion2 or later
```

## License

rewavi is licensed under the [GNU Genenral Public License version 2 or later](LICENSE.TXT).


CHANGELOG.md
# Changelog

### Version 0.06 (2018-01-09)

- Changed progress output interval.

### Version 0.05 (2018-01-09)

- Refactored common code.

### Version 0.04 (2018-01-08)

- Fixed progress output.

### Version 0.03 (2018-01-08)

- Added progress output.
- Added print output format.
- Refactored argv parsing.
- Added channel mask argv validation.
- Added check for invalid argv.

### Version 0.02 (2018-01-06)

- Added Visual Studio 2017 support.
- Added x64 build configuration.
- Added Cake build script.
- Add appveyor configuration for using Cake build system.
- Bug fixes.

### Version 0.01 (2011-11-19)

- Initial release.


src/rewavi.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdarg.h>
#include <io.h>
#include <fcntl.h>

#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <vfw.h>

#ifndef WAVE_FORMAT_IEEE_FLOAT
#define WAVE_FORMAT_IEEE_FLOAT 3
#endif

#include "version.h"

#pragma comment(lib, "vfw32.lib")

#define SUBFORMAT_GUID(fmt) \
    {fmt, 0x0000, 0x0010, {0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}}

#define DEFAULT_MASK_COUNT 9
#define DEFAULT_MASK {0, 4, 3, 259, 51, 55, 63, 319, 255}
#define BUFFSIZE 0x20000 /* 128KiB */

typedef enum {
    FORMAT_WAV,
    FORMAT_RAW
} format_t;

static void usage(void);
static void usage2(void);
static uint32_t numofbits(uint32_t bits);

int main(int argc, char **argv)
{
    if (argc < 2)
    {
        usage();
        return 1;
    }

    if (!strcmp(argv[1], "-h"))
    {
        usage2();
        return 1;
    }

    if (!argv[2] || !_stricmp(argv[2], "-x") || !_stricmp(argv[2], "-r"))
    {
        fprintf(stderr, "output target is not specified.\n");
        return 2;
    }

    char* input = argv[1];
    char* output = argv[2];

    format_t format = FORMAT_WAV;
    if (argc == 4)
    {
        if (!_stricmp(argv[3], "-r"))
        {
            format = FORMAT_RAW;
        }
        else if (!_stricmp(argv[3], "-x"))
        {
            fprintf(stderr, "channel mask is not specified.\n");
            return 2;
        }
        else
        {
            fprintf(stderr, "raw format is not specified.\n");
            return 2;
        }
    }

    uint32_t chmask = 0;
    if (argc == 5)
    {
        if (!_stricmp(argv[3], "-x"))
        {
            chmask = atoi(argv[4]);
        }
        else
        {
            fprintf(stderr, "channel mask is not specified.\n");
            return 2;
        }
    }

    AVIFileInit();
    FILE* output_fh = NULL;
    PAVIFILE avifile = NULL;
    PAVISTREAM avistream = NULL;
    AVISTREAMINFO stream_info;
    WAVEFORMATEX wavefmt;
    LONG fmtsize = sizeof(WAVEFORMATEX);
    int id = 0;

    if (AVIFileOpen(&avifile, input, OF_READ | OF_SHARE_DENY_WRITE, NULL))
    {
        fprintf(stderr, "Could not open AVI file \"%s\".\n", input);
        AVIFileExit();
        return 1;
    }

    while (1)
    {
        if (AVIFileGetStream(avifile, &avistream, 0, id))
            break;

        if (AVIStreamInfo(avistream, &stream_info, sizeof(AVISTREAMINFO)))
            goto release_stream;

        if (stream_info.fccType != streamtypeAUDIO)
            goto release_stream;

        fprintf(stderr, "Found an audio track(ID %d) in %s\n", id, input);
        if (AVIStreamReadFormat(avistream, 0, &wavefmt, &fmtsize))
            goto release_stream;

        if (wavefmt.wFormatTag == WAVE_FORMAT_PCM ||
            wavefmt.wFormatTag == WAVE_FORMAT_IEEE_FLOAT)
            break;

    release_stream:
        AVIStreamRelease(avistream);
        avistream = NULL;
        id++;
    }

    if (!avistream)
    {
        fprintf(stderr, "Could not find PCM audio track in \"%s\".\n", input);
        AVIFileRelease(avifile);
        AVIFileExit();
        return 1;
    }

    fprintf(stderr, "streamID %d: %uchannels, %uHz, %ubits, %.3fseconds.\n",
        id, wavefmt.nChannels, wavefmt.nSamplesPerSec, wavefmt.wBitsPerSample,
        1.0 * stream_info.dwLength * stream_info.dwScale / stream_info.dwRate);

    if (wavefmt.wFormatTag == WAVE_FORMAT_IEEE_FLOAT)
        fprintf(stderr, "Audio track contains floating-point samples.\n");

    if (BUFFSIZE < wavefmt.nBlockAlign)
    {
        fprintf(stderr, "Too large samples; audio track is definitely invalid.\n");
        AVIStreamRelease(avistream);
        AVIFileRelease(avifile);
        AVIFileExit();
        return 1;
    }

    if (chmask)
    {
        if (numofbits(chmask) != wavefmt.nChannels)
        {
            fprintf(stderr, "Invalid channel mask was specified.\n");
            AVIStreamRelease(avistream);
            AVIFileRelease(avifile);
            AVIFileExit();
            return 1;
        }
    }
    else
    {
        uint32_t default_chmask[DEFAULT_MASK_COUNT] = DEFAULT_MASK;
        chmask = wavefmt.nChannels < DEFAULT_MASK_COUNT ?
            default_chmask[wavefmt.nChannels] : (1 << wavefmt.nChannels) - 1;
    }

    int dupout = 0;
    if (!strcmp(output, "-"))
    {
        dupout = _dup(_fileno(stdout));
        fclose(stdout);
        _setmode(dupout, _O_BINARY);
        output_fh = _fdopen(dupout, "wb");
    }
    else
    {
        fopen_s(&output_fh, output, "wb");
    }

    if (!output_fh)
    {
        fprintf(stderr, "Fail to create/open file.\n");
        AVIStreamRelease(avistream);
        AVIFileRelease(avifile);
        AVIFileExit();
        return 1;
    }

    char buffer[BUFFSIZE];
    uint32_t samples_in_buffer = BUFFSIZE / wavefmt.nBlockAlign;
    LONG samples_read;
    DWORD nextsample = 0;

    if (format == FORMAT_WAV)
    {
        uint32_t headersize = chmask ? 60 : 36;
        uint64_t filesize = (uint64_t)wavefmt.nBlockAlign * stream_info.dwLength;
        if (filesize > 0xFFFFFFFF - headersize)
        {
            fprintf(stderr, "WAV file will be larger than 4GB.\n");
            filesize = wavefmt.nBlockAlign * (0xFFFFFFFF - headersize / wavefmt.nBlockAlign);
        }
        uint32_t size_for_write = (uint32_t)filesize + headersize;

        /* riff file header */
        fwrite("RIFF", 1, 4, output_fh);
        fwrite(&size_for_write, 4, 1, output_fh);
        fwrite("WAVE", 1, 4, output_fh);

        /* format chunk */
        fwrite("fmt ", 1, 4, output_fh);
        size_for_write = chmask ? 40 : 16;
        fwrite(&size_for_write, 4, 1, output_fh);

        /* As for the WAVEFORMATEX / WAVEFORMATEXTENSIBLE structure, alignment is taken into consideration. */
        if (!chmask)
        {
            /* never write cbSize! */
            fwrite(&wavefmt, fmtsize - 2, 1, output_fh);
        }
        else
        {
            WAVEFORMATEXTENSIBLE wavefmt_ext = {
                wavefmt, {wavefmt.wBitsPerSample}, chmask,
                SUBFORMAT_GUID(wavefmt.wFormatTag) };
            wavefmt_ext.Format.wFormatTag = WAVE_FORMAT_EXTENSIBLE;
            wavefmt_ext.Format.cbSize = 22;
            fwrite(&wavefmt_ext, sizeof(WAVEFORMATEXTENSIBLE), 1, output_fh);
        }

        /* data chunk*/
        fwrite("data", 1, 4, output_fh);
        size_for_write = (uint32_t)filesize;
        fwrite(&size_for_write, 4, 1, output_fh);

        fflush(output_fh);
    }

    fprintf(stderr, "Writing %s file %s ...\n", (format == FORMAT_WAV) ? "WAV" : "RAW", dupout ? "to stdout" : output);

    /* fraction processing at first. */
    AVIStreamRead(avistream, 0, stream_info.dwLength % samples_in_buffer, &buffer, BUFFSIZE, NULL, &samples_read);
    while (nextsample < stream_info.dwLength)
    {
        fwrite(buffer, wavefmt.nBlockAlign, samples_read, output_fh);
        nextsample += samples_read;
        fprintf(stderr, "\rProgress: %d%% (%lu/%lu)", 
            (int)(((double)nextsample / (double)stream_info.dwLength) * 100), 
            nextsample, stream_info.dwLength);
        /* The last reading is just ignored. */
        AVIStreamRead(avistream, nextsample, samples_in_buffer, &buffer, BUFFSIZE, NULL, &samples_read);
    }

    fprintf(stderr, "\n");
    fflush(output_fh);

    fclose(output_fh);
    AVIStreamRelease(avistream);
    AVIFileRelease(avifile);
    AVIFileExit();

    if (nextsample != stream_info.dwLength)
    {
        fprintf(stderr, "Writing failed.\n");
        return 1;
    }
    else
    {
        fprintf(stderr, "File written successfully.\n");
        return 0;
    }
}

static void usage(void)
{
    fprintf(stderr,
        "REWAVI - rewritten/modified WAVI ver.%s\n"
        " (c) 2011 Oka Motofumi\n"
        " (c) 2018 Wieslaw Soltes\n"
        " WAVI original - (c) 2007 Tamas Kurucsai, with lots of help from tebasuna51\n"
        "Licensed under the terms of the GNU General Public License version2 or later.\n"
        "\n"
        "Usage: rewavi <input> <output> [ -r | -x [<mask>] ]\n"
        "  input  : input AVI1.0( or AviSynth Script)file name.\n"
        "  output : output WAV file name. '-' means output to stdout\n"
        "  -r : Write a raw file of samples without the WAV header.\n"
        "  -x : Write an extended WAV header containing channel mask for\n"
        "       multi-channel audio.\n"
        "       If channel mask is not specified, the defaut value will be set.\n"
        "\n"
        " 'rewavi -h' will display more complicated description.\n", VERSION);
}

static void usage2(void)
{
    usage();
    fprintf(stderr,
        "\n"
        "This utility extracts the first uncompressed PCM audio track\n"
        "from an AVI1.0 file and saves it to a WAV file. This is not quite\n"
        "useful for most AVI files since they usually contain some kind\n"
        "of compressed audio, but it can come handy when it's needed to\n"
        "save an audio track from AviSynth.\n"
        "\n"
        "If <input> is a valid AVI1.0 file which contains PCM audio,\n"
        "the audio track will be written to <output> as a WAV(or RAW) file.\n"
        "If '-' is passed as <output>, the WAV file will be written\n"
        "to the standard output.\n"
        "If <output> is not given, only information will be printed\n"
        "about the audio track.\n"
        "\n"
        "On success, the exit code will be 0 and the first line printed\n"
        "to the standard error will look like the following:\n"
        "    streamID <n>: <c>channels, <r>Hz, <b>bits, <l>sec.\n"
        "\n"
        "Where <n> is IDnumber of audio track that comtaining PCM audio, <c> is\n"
        "the number of audio channels, <r> is the sampling rate, <b> is the number\n"
        "of bits per sample and <l> is the length of the track in seconds.\n"
        "\n"
        "If the audio track contains floating-point samples, the next line\n"
        "printed to the standard error will be:\n"
        "Audio track contains floating-point samples.\n"
        "\n"
        "If an error occurs, the exit code will be 1 and some useful error\n"
        "message will be printed to the standard error.\n"
        "\n"
        "WAV files larger than 4 GB may be created. However, such WAV files\n"
        "are non-standard and may not be handled correctly by some players\n"
        "and encoders. A warning will be printed to the standard error when\n"
        "such a WAV file is created.\n"
        "\n"
        "The default channel masks are:\n"
        "\n"
        "Mask  Ch.  MS channels                   Description\n"
        "----  ---  ----------------------------  ----------------\n"
        "   4   1                       FC        Mono\n"
        "   3   2                          FR FL  Stereo\n"
        " 259   3   BC                     FR FL  First Surround\n"
        "  51   4              BR BL       FR FL  Quadro\n"
        "  55   5              BR BL    FC FR FL  like Dpl II (without LFE)\n"
        "  63   6              BR BL LF FC FR FL  Standard Surround\n"
        " 319   7   BC         BR BL LF FC FR FL  With back center\n"
        " 255   8      FLC FRC BR BL LF FC FR FL  With front center left/right\n"
        "\n"
        "Some other common channel masks:\n"
        "\n"
        "Mask  Ch. MS channels                    Description\n"
        "----  --- -----------------------------  ----------------\n"
        "   7   3                       FC FR FL\n"
        " 263   4  BC                   FC FR FL  like Dpl I\n"
        " 271   5  BC                LF FC FR FL\n"
        "  59   5              BR BL LF    FR FL\n");
}

static uint32_t numofbits(uint32_t bits)
{
    bits = (bits & 0x55555555) + (bits >> 1 & 0x55555555);
    bits = (bits & 0x33333333) + (bits >> 2 & 0x33333333);
    bits = (bits & 0x0f0f0f0f) + (bits >> 4 & 0x0f0f0f0f);
    bits = (bits & 0x00ff00ff) + (bits >> 8 & 0x00ff00ff);
    return (bits & 0x0000ffff) + (bits >> 16 & 0x0000ffff);
}


src/resilence.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdarg.h>
#include <io.h>
#include <fcntl.h>

#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <vfw.h>

#ifndef WAVE_FORMAT_IEEE_FLOAT
#define WAVE_FORMAT_IEEE_FLOAT 3
#endif

#include "version.h"

#define DEFAULT_LENGTH 1.0
#define DEFAULT_CHANNELS 2
#define DEFAULT_SAMPLERATE 48000
#define DEFAULT_SAMPLEBITS 16
#define NO_DEFAULT 0
#define BUFFSIZE 0x100000 /* 1MiB*/

static void usage(void);

int main(int argc, char **argv)
{
    if (argc < 2)
    {
        usage();
        return 1;
    }

    FILE *output_fh = NULL;

    if (!strcmp(argv[1], "-f") || !strcmp(argv[1], "-l") ||
        !strcmp(argv[1], "-c") || !strcmp(argv[1], "-b"))
    {
        fprintf(stderr, "Missing output target.\n");
        return 1;
    }

    WAVEFORMATEX wavefmt = { WAVE_FORMAT_PCM, DEFAULT_CHANNELS, DEFAULT_SAMPLERATE,
                             NO_DEFAULT, NO_DEFAULT, DEFAULT_SAMPLEBITS, 0 };
    double length = DEFAULT_LENGTH;
    int samplebits_given = 0;

    int i = 1;
    while (argv[++i])
    {
        if (argv[i][0] != '-')
        {
            fprintf(stderr, "Invarid argument '%s'\n", argv[i]);
            return 1;
        }
        switch (argv[i][1])
        {
        case 'f':
            wavefmt.wFormatTag = WAVE_FORMAT_IEEE_FLOAT;
            wavefmt.wBitsPerSample = 32;
            break;
        case 'l':
            length = atof(argv[++i]);
            if (length <= 0)
            {
                fprintf(stderr, "Invalid length given after -l\n");
                return 1;
            }
            break;
        case 'c':
            wavefmt.nChannels = atoi(argv[++i]);
            if (!wavefmt.nChannels || wavefmt.nChannels > 32)
            {
                fprintf(stderr, "Invalid number of channels '%s' given after -c\n", argv[i]);
                return 1;
            }
            break;
        case 's':
            wavefmt.nSamplesPerSec = atoi(argv[++i]);
            if (!wavefmt.nSamplesPerSec || wavefmt.nSamplesPerSec > 192000)
            {
                fprintf(stderr, "Invalid sample rate '%s' given after -s\n", argv[i]);
                return 1;
            }
            break;
        case 'b':
            wavefmt.wBitsPerSample = atoi(argv[++i]);
            if (!wavefmt.wBitsPerSample || wavefmt.wBitsPerSample % 8 || wavefmt.wBitsPerSample > 32)
            {
                fprintf(stderr, "Invalid bits per sample '%s' given after -b\n", argv[i]);
                return 1;
            }
            samplebits_given = 1;
            break;
        default:
            fprintf(stderr, "Invarid argument '%s'\n", argv[i]);
            return 1;
        }
    }

    if (wavefmt.wFormatTag + samplebits_given > 3)
    {
        fprintf(stderr, "'-f' and '-b' cannot be used at the same time.\n");
        return 1;
    }

    wavefmt.nBlockAlign = wavefmt.nChannels * wavefmt.wBitsPerSample >> 3;
    wavefmt.nAvgBytesPerSec = wavefmt.nBlockAlign * wavefmt.nSamplesPerSec;
    uint64_t filesize = (uint64_t)(wavefmt.nAvgBytesPerSec * length);
    uint32_t size_for_write = (uint32_t)filesize + 36;
    if (filesize > 0xFFFFFFFF - 36)
    {
        fprintf(stderr, "WAV file will be larger than 4 GB.\n");
        size_for_write = wavefmt.nBlockAlign * ((0xFFFFFFFF - 36) / wavefmt.nBlockAlign) + 36;
    }

    int dupout = 0;
    if (!strcmp(argv[1], "-"))
    {
        dupout = _dup(_fileno(stdout));
        fclose(stdout);
        _setmode(dupout, _O_BINARY);
        output_fh = _fdopen(dupout, "wb");
    }
    else
    {
        fopen_s(&output_fh, argv[1], "wb");
    }

    if (!output_fh)
    {
        fprintf(stderr, "Fail to create/open file.\n");
        return 1;
    }

    fprintf(stderr, "Writing silence WAV file %s ...\n", dupout ? "to stdout" : argv[1]);

    fwrite("RIFF", 1, 4, output_fh);
    fwrite(&size_for_write, 4, 1, output_fh);
    fwrite("WAVE", 1, 4, output_fh);
    fwrite("fmt ", 1, 4, output_fh);
    fwrite("\x10\x00\x00\x00", 1, 4, output_fh);
    /* never write cbSize! */
    fwrite(&wavefmt, sizeof(WAVEFORMATEX) - 2, 1, output_fh);
    fwrite("data", 1, 4, output_fh);
    size_for_write -= 36;
    fwrite(&size_for_write, 4, 1, output_fh);

    fflush(output_fh);

    char data[BUFFSIZE] = { 0 };
    uint64_t wrote = filesize % BUFFSIZE;
    fwrite(data, 1, (size_t)wrote, output_fh);
    while (wrote < filesize)
    {
        fwrite(data, 1, BUFFSIZE, output_fh);
        wrote += BUFFSIZE;
    }

    fflush(output_fh);
    fclose(output_fh);

    if (wrote != filesize)
    {
        fprintf(stderr, "Writing failed...\n");
        return 1;
    }
    else
    {
        fprintf(stderr, "WAV file written successfully.\n");
        return 0;
    }
}

static void usage(void)
{
    fprintf(stderr,
        "RESILENCE - rewritten/modified SILENCE ver.%s\n"
        "   (c) 2011 Oka Motofumi\n"
        "   (c) 2018 Wieslaw Soltes\n"
        "   original SILENCE (c) 2007 Tamas Kurucsai\n"
        " Licensed under the terms of the GNU General Public License verion2 or later\n"
        "\n"
        "This tool create silent WAV files.\n"
        "Usage: resilence <output> [-l <length>] [-c <channels>] [-s <sample-rate>]\n"
        "                          [-b <bits>] [-f]\n"
        "\n"
        "  output: output WAV file name. \"-\" means output to stdout.\n"
        "  -l : length of output stream. (unit:sec  default=1.0)\n"
        "  -c : number of channels of output stream. (default:2  min:1  max:32)\n"
        "  -s : sample rate of output stream. (unit:Hz  default:48000 max:192000)\n"
        "  -b : bits-per-sample of output stream. this value must be 8, 16, 24,\n"
        "       or 32. (default:16)\n"
        "  -f : use 32bit floating-point samples instead of 8/16/24/32bit integer.\n"
        "       this option cannot be used with -b at the same time.\n",
        VERSION);
}
